### DISCUSSION TITLE
DNS peer autodiscovery stopped working in 3.12 -> 3.13 upgrade

### DISCUSSION THREAD

**Discussion by bkienker** [rabbitmq/rabbitmq-server#12151] [Questions] [answered]
I have been running a 3-node RabbitMQ cluster with the DNS autodiscovery cluster formation backend for a while now. I recently attempted to upgrade from 3.12.14 to 3.13.1 and found that the nodes no longer form a cluster. I attempted the upgrade in a development environment by stopping and removing all existing instances except for their DNS records, then starting one node, letting it boot, and starting the remaining two nodes.

To begin with, I did not make any modifications to the rabbitmq.conf or any port mappings to the containers running RabbitMQ. Having gone through the release notes and latest docs, I do not see where any changes to the DNS peer autodiscovery configuration would cause the feature to stop working between 3.12 and 3.13.

I am stumped because it appears I am the only one having this issue. I feel confident it is something to do with the 3.12 -> 3.13 upgrade because if I revert the RabbitMQ version to 3.12.14 with the exact same configuration, the cluster forms exactly the way it always has. I've enabled debug logging on multiple nodes, and I have not observed any errors in the DNS autodiscovery process, or really any errors at all.

Were there any changes to DNS autodiscovery in version 3.13 that would prevent a previously-working cluster from peering correctly? Thanks in advance for any advice you can offer.

**Comment by lukebakken** [comment]
Please re-try using version 3.13.7. Thank you.

**Reply by bkienker** [reply]
Thank you for the extremely quick reply. Unfortunately, I am observing the same behavior with version 3.13.7. Starting node 1 is normal, but node 2 does not discover node 1 on startup and forms its own cluster.

**Comment by lukebakken** [comment]
In your environment, if you create a _FRESH_ 3.13.7 cluster using DNS discovery, does it succeed?

**Reply by bkienker** [reply]
It should be a blank slate each time -- I'm running RabbitMQ in containers, so when I reinitialize the cluster, my process is this:

1. Stop all RabbitMQ containers
2. Clear all temporary data from disk with `podman system reset`
3. Start node 1, wait for the rabbit app to come up
4. Start nodes 2/3 and wait for the rabbit app to come up

The pod resource YAML, rabbitmq.conf, definitions.json, and DNS Records remain the same, but all original cluster data should be wiped in each iteration.

**Reply by lukebakken** [reply]
OK, so you're not really doing an upgrade (i.e. with the old data in place). You should be able to provide a git repository I can clone and then run a `podman` command (or `docker compose`, I'm familiar with that), to demonstrate the issue.

**Reply by bkienker** [reply]
I'm not sure that I will be able to provide something truly portable. My configuration management handles deploying the necessary Pod resource files and configurations across three virtual machines, which each run a separate RabbitMQ container that clusters based on the local hostname/IP/DNS record, all of which are specific to my domain.

Would it be helpful for me to send a single copy of each of the files required to run one node? If so, I can make generic copies of each, but it will still require you still fill in a fair number of blanks and configure DNS records, etc.

**Reply by lukebakken** [reply]
https://www.rabbitmq.com/docs/cluster-formation#dns-peer-discovery-overview

In your environment, do the following (please provide the full transcript of the command you run, and its output):

* Start your RabbitMQ nodes, with or without RabbitMQ running.
* On each node, do a DNS lookup for your `cluster_formation.dns.hostname` host. 
* Then, do a reverse DNS lookup on each of the provided IP addresses, which should resolve to the _FULL_ host names for your RabbitMQ nodes.

**Reply by bkienker** [reply]
Thank you, here's that information. The startup output for each RMQ node is attached as a file, for readability. Please let me know if you need anything else and I will provide it ASAP.

Here's the DNS bits:
<details><summary>01</summary>

```
service_account@rabbitmq-01 $ grep cluster_formation.dns.hostname .rabbitmq/rabbitmq.conf
cluster_formation.dns.hostname = rabbitmq-dev.subdomain.mydomain.net
service_account@rabbitmq-01 $ dig +short rabbitmq-dev.subdomain.mydomain.net
10.0.0.39
10.0.0.47
10.0.0.40
service_account@rabbitmq-01 $ dig +short -x 10.0.0.39
rabbitmq-01.mydomain.net.
service_account@rabbitmq-01 $ dig +short -x 10.0.0.40
rabbitmq-02.mydomain.net.
service_account@rabbitmq-01 $ dig +short -x 10.0.0.47
rabbitmq-03.mydomain.net.
```
</details>

<details><summary>02</summary>

```
service_account@rabbitmq-02 $ grep cluster_formation.dns.hostname .rabbitmq/rabbitmq.conf
cluster_formation.dns.hostname = rabbitmq-dev.subdomain.mydomain.net
service_account@rabbitmq-02 $ dig +short rabbitmq-dev.subdomain.mydomain.net
10.0.0.47
10.0.0.39
10.0.0.40
service_account@rabbitmq-02 $ dig +short -x 10.0.0.39
rabbitmq-01.mydomain.net.
service_account@rabbitmq-02 $ dig +short -x 10.0.0.40
rabbitmq-02.mydomain.net.
service_account@rabbitmq-02 $ dig +short -x 10.0.0.47
rabbitmq-03.mydomain.net.
```
</details>

<details><summary>03</summary>

```
service_account@rabbitmq-03 $ grep cluster_formation.dns.hostname .rabbitmq/rabbitmq.conf
cluster_formation.dns.hostname = rabbitmq-dev.subdomain.mydomain.net
service_account@rabbitmq-03 $ dig +short rabbitmq-dev.subdomain.mydomain.net
10.0.0.40
10.0.0.47
10.0.0.39
service_account@rabbitmq-03 $ dig +short -x 10.0.0.39
rabbitmq-01.mydomain.net.
service_account@rabbitmq-03 $ dig +short -x 10.0.0.40
rabbitmq-02.mydomain.net.
service_account@rabbitmq-03 $ dig +short -x 10.0.0.47
rabbitmq-03.mydomain.net.
```
</details>

[rabbitmq-03.txt](https://github.com/user-attachments/files/16790082/rabbitmq-03.txt)
[rabbitmq-02.txt](https://github.com/user-attachments/files/16790083/rabbitmq-02.txt)
[rabbitmq-01.txt](https://github.com/user-attachments/files/16790084/rabbitmq-01.txt)

**Reply by lukebakken** [reply]
The reverse DNS lookups on `rabbitmq-03` don't appear to return the expected hostnames - is there a mistake in what you pasted, or is this an actual issue?

**Reply by bkienker** [reply]
That's a copy/paste mistake, apologies, I will fix it.

**Reply by lukebakken** [reply]
I'll see how we test this part of RabbitMQ, and may be able to find time to investigate soon-ish.

**Reply by bkienker** [reply]
Thank you for your help!

**Comment by lukebakken** [comment]
@bkienker from your 3.13 environment, can you start up RabbitMQ, and send us the logs? DNS peer discovery should log lines with `Addresses discovered via` in them. You can check your 3.12 environment for these log entries as well.

**Reply by michaelklishin** [reply]
Logs [at debug level](https://www.rabbitmq.com/docs/logging#debug-logging) to be more specific.

**Comment by bkienker** [comment]
Thank you both, I have started each cluster node from a blank slate and attached all logs. I'm not seeing the `Addresses discovered via` line. Nothing else is really standing out to me. Please let me know what additional information I can provide to assist.

[rabbitmq-03.log](https://github.com/user-attachments/files/16895005/rabbitmq-03.log)
[rabbitmq-02.log](https://github.com/user-attachments/files/16895006/rabbitmq-02.log)
[rabbitmq-01.log](https://github.com/user-attachments/files/16895007/rabbitmq-01.log)

**Reply by lukebakken** [reply]
Are you sure it's a blank state? There is no indication that peer discovery/clustering is even being attempted in those log files.

**Reply by lukebakken** [reply]
I'm getting a local env set up to test. I'm not using k8s

**Reply by bkienker** [reply]
I've attempted to get as blank a slate as possible. Attached are logs from 01 and 02, now running the official image instead of the Bitnami one, after running `podman system reset`. They seem to be logging much more about DNS peer discovery, but `rabbitmqctl cluster_status` still shows that both nodes form a cluster only containing themselves.

[rmq-01.log](https://github.com/user-attachments/files/16897917/rmq-01.log)
[rmq-02.log](https://github.com/user-attachments/files/16897918/rmq-02.log)

**Reply by lukebakken** [reply, marked as answer]
Yep, that is more useful. Here are the relevant log lines from `rabbitmq-01`:

```
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.331972+00:00 [info] <0.253.0> DB: virgin node -> run peer discovery
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.364934+00:00 [info] <0.253.0> Addresses discovered via A records of rabbitmq-dev.mydomain.net: 10.90.70.47, 10.90.70.40, 10.90.70.39
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.368992+00:00 [info] <0.253.0> Addresses discovered via AAAA records of rabbitmq-dev.mydomain.net:
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.369059+00:00 [debug] <0.253.0> Peer discovery: backend returned the following configuration:
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.369059+00:00 [debug] <0.253.0>   {ok,{['rabbit@rabbitmq-03','rabbit@rabbitmq-02',rabbit@host],
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.369059+00:00 [debug] <0.253.0>        disc}}
...
...
...
Sep 05 12:43:42 rabbitmq-01 rabbitmq[4071336]: 2024-09-05 16:43:42.629572+00:00 [info] <0.253.0> Peer discovery: discovered no peer nodes to cluster with. Some discovery backends can filter nodes out based on a readiness criteria. Enabling debug logging might help troubleshoot.
```

Notice that the reverse DNS lookups on those three IP addresses returned the following host names:
* `rabbitmq-03`
* `rabbitmq-02`
* `host`

Obviously, the last one is not correct.

On node `rabbitmq-01`, can you please do the following?

* A reverse DNS on the IP address (`dig -x`) for that node to see what is returned?
* Run the following `erl` command, using that node's IP (replace the `X`):
    ```
    erl -noinput -eval 'io:format("~tp~n",[inet:gethostbyaddr("10.90.70.X")]),init:stop().'
    ```

**Reply by bkienker** [reply]
That is very odd. Here's the `dig` from the host VM:
```
service_account@rabbitmq-01 $ dig +noall +answer -x 10.90.70.39
39.70.90.10.in-addr.arpa. 300	IN	PTR	rabbitmq-01.mydomain.net.
```

From the RabbitMQ container shell:
```
root@rabbitmq-01:/# erl -noinput -eval 'io:format("~tp~n",[inet:gethostbyaddr("10.90.70.39")]),init:stop().'
{ok,{hostent,"host.containers.internal",[],inet,4,[{10,90,70,39}]}}
```

Comparing to the same DNS query for the 02 node's IP address:
```
root@rabbitmq-01:/# erl -noinput -eval 'io:format("~tp~n",[inet:gethostbyaddr("10.90.70.40")]),init:stop().'
{ok,{hostent,"rabbitmq-02.mydomain.net",[],inet,4,[{10,90,70,40}]}}
```

**I've found the culprit** -- there's an entry in the container's `/etc/hosts` file that is superseding the DNS result:
```
root@rabbitmq-01:/# grep host.containers.internal /etc/hosts
10.90.70.39	host.containers.internal
```

After removing that line, the DNS query is returning the correct FQDN:
```
root@rabbitmq-01:/# erl -noinput -eval 'io:format("~tp~n",[inet:gethostbyaddr("10.90.70.39")]),init:stop().'
{ok,{hostent,"rabbitmq-01.mydomain.net",[],inet,4,[{10,90,70,39}]}}
```

Thank you for the assistance. I will see if I can stop podman from "helping" me with the /etc/hosts file. I apologize for the wild goose chase.

**Reply by michaelklishin** [reply]
@bkienker this sounds like something we must [document](https://github.com/rabbitmq/rabbitmq-website/blob/main/versioned_docs/version-3.13/cluster-formation.md).

Also note that Erlang [allows you to tweak DNS resolution a great deal](https://www.erlang.org/doc/apps/erts/inet_cfg.html) and likely bypass `/etc/hosts` entirely.

**Reply by lukebakken** [reply]
> I apologize for the wild goose chase.

No worries! I'm still mystified as to how this works fine with RabbitMQ 3.12 since the DNS lookup code hasn't changed. What version of Erlang are you using in your 3.12 environments?

**Reply by michaelklishin** [reply]
Peer discovery did change in 3.13, so its place in the node boot sequence may matter. For example, with 3.12 it could have happened before Podman had a chance to mess with the `/etc/hosts` file.

**Reply by bkienker** [reply]
I had been trying out the last minor 3.12 container image, looks like 3.12.14, which has this version:
```
root@da0aff08909a:/# erl --version
Erlang/OTP 25 [erts-13.2.2.10] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:1] [jit:ns]

Eshell V13.2.2.10  (abort with ^G)
```

Inspecting some of my older clusters, they appear to have the /etc/hosts line present, so whatever change took place appears to have it finding the /etc/hosts entry before/instead of the DNS PTR record.

Fortunately, it's relatively easy to have podman stop adding the entry to /etc/hosts. I don't use any of the podman internal networking, so I don't have any use for the entry anyway. See https://github.com/containers/common/blob/main/docs/containers.conf.5.md:
```
host_containers_internal_ip=""

Set the ip for the host.containers.internal entry in the containers /etc/hosts file. This can be set to "none" to disable adding this entry. By default it will automatically choose the host ip.
```

After setting that flag to `none`, the Erlang process is getting the correct PTR record for the host IP and clustering is working properly.

**Reply by michaelklishin** [reply]
@bkienker thank you for the details, I will try to reflect this in the cluster formation docs.

**Reply by michaelklishin** [reply]
@bkienker here is a doc update which should at least guide the reader in the right direction https://github.com/rabbitmq/rabbitmq-website/commit/5baa01ed2512d632e448197e9d4b941287562f24.

Sadly Podman documentation, even the guide dedicated to networking, does not mention `host_containers_internal_ip` so I had to link to the `containers.conf(5)` source document above.

Thank you for sharing your findings.

**Reply by bkienker** [reply]
This is fantastic, y'all are super on top of things. That containers.conf link should cover the vast majority of cases with podman, as containerd is the default container engine on RHEL-flavor distros these days. That addition to the RabbitMQ docs should be very helpful to anyone running into the same problem in the future. Thank you both again for your help!
